design.md

```diff
@@ -1,2 +1,97 @@
+
+When created, the load balancer would create an internal map of each host to the number of connections the host is currently handling.
+
+When the server gets a TCP connection, it will pass it to the request forwarding library to route it to the appropriate upstream:
+
+```go
+func (lb *LoadBalancer) HandleConnection(ctx context.Context, conn net.Conn)
+```
+
+At a high level, the `HandleConnection` function will:
+
+- Iterate over each of the hosts to find the one with the fewest connections
+    - The counters used for connections will be atomic (e.g. protected from concurrent access with a mutex, or use `sync/atomic` functions) to avoid race conditions
+- Dial the host to initiate a TCP connection with it
+    - If the connection can't be established, write an error message to the `conn` passed in. Close the connection and return without further processing.
+    - To avoid deadlocks, we'll use a timeout here to bail if the host doesn't respond in t seconds.
+- Increment (again, thread safely) the connection count for the host
+- Defer decrementing the connection count for the host
+- In two separate goroutines, `io.Copy` the connection passed in to the connection to the upstream and vice versa
+- Block for each goroutine to signal its completion
+- After completion, close both connections
+
+### Client Rate Limiter
+
+This library will implement a simple, in-memory rate limiter for client requests. Users will create a rate limiter by passing in a map of client identifier -> # of requests allowed per second like so:
+
+```go
+func NewRateLimiter(requestsPerClient map[string]int) *RateLimiter
+```
+
+(To keep things simple, we'll only allow per-second limits rather than customizing the timeframe used)
+
+Consumers of the library can find out whether a request is allowed using the `isRequestAllowed` function:
+
+```go
+func (rl *RateLimiter) IsRequestAllowed(ctx context.Context, clientId string) (bool, error)
+```
+
+Internally the library will use [the token bucket algorithm](https://en.wikipedia.org/wiki/Token_bucket) to rate limit requests. At creation it will construct a bucket for each client. Calling `IsRequestAllowed` will trigger the library to query the corresponding bucket and determine whether the client has enough tokens (always 1 for simplicity's sake) to make the request.
+
+The bucket will use a mutex to synchronize access to its token count to prevent multithreading issues where quick requests from the same client don't accurately read/modify the value. The rate the bucket refills will be hard coded to keep things simple, though in reality this would be configurable.
+
+## Server
+
+The server is responsible for authentication and authorization before handing off the main work to the load balancing library. Authentication will be done with mTLS, with certificates stored in the repo for the sake of simplicity. Authorization will also be hard coded; we'll simply keep a struct of which upstreams are available to which clients.
```

Comment:
Good question. I added some additional detail here; basically I plan to use the email in the client certs' SAN as a "user" we can check our hard-coded list against to see if that user is authorized to access the TCP server on a given port. (In reality we'd probably want to use our own SAN extension, not just email)

Let me know if that clears it up - it's late and my brain is fried so it may be total nonsense :sweat_smile: